Add: Allow Mudlet to be controlled by Claude / ChatGPT

vadi2 opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Adds an MCP server to Mudlet, allowing LLM's to talk to Mudlet. This enables unexperienced scripters to have an AI write code straight to Mudlet for them.
#### Motivation for adding to Mudlet
Fix https://github.com/Mudlet/Mudlet/issues/8212
#### Other info (issues closed, discussion etc)
Better first-timer friendliness.

[Screencast from 2025-09-19 17-58-33.webm](https://github.com/user-attachments/assets/6db93141-b33e-4fef-971e-06021d96ac8d)

vadi2 committed:
Initial version

vadi2 committed:
Initial version with support for listing tools

vadi2 committed:
Make MCP server an optional compile-time feature

vadi2 committed:
Improve spec compliance

vadi2 committed:
Remove spammy message

vadi2 committed:
Slim down all tools to just one - Lua

vadi2 committed:
Implement profile parameter and simplify responses

vadi2 committed:
Merge development into add-mcp-server

Resolved conflicts in:
- src/Host.cpp: Integrated MCP server initialization with development changes
- src/Host.h: Added MCP server pointer declaration
- src/dlgProfilePreferences.cpp: Added MCP server protocol option alongside MNES

vadi2 committed:
Add missing newlines at end of files

vadi2 committed:
Add Qt6 HttpServer dependency to CI workflows

Adds qthttpserver module to all CI workflows to support the MCP server feature.

vadi2 committed:
Add Qt6 WebSockets dependency for HttpServer

Qt6::HttpServer requires Qt6::WebSockets as a dependency.

vadi2 committed:
Merge branch 'development' into add-mcp-server

vadi2 committed:
Merge remote-tracking branch 'origin/development' into add-mcp-server

# Conflicts:
#	src/Host.h

vadi2 committed:
Fix CMake target name for MCP server compile definition

vadi2 committed:
Remove non-existent pre_guard.h and post_guard.h includes

mpconley commented:
## MCP Server Testing Results ✅

Just successfully tested the MCP server functionality! Here's what we experienced:

**Setup Process:**
1. **Enabled MCP server in Mudlet**: Went to Settings → Profile Preferences → General → Protocols → checked "MCP: Model Context Protocol Server"
2. **Added Mudlet to VS Code MCP config**: Added `"mudlet": {"type": "http", "url": "http://localhost:11235/"}` to the mcp.json configuration
3. **Successfully executed commands**: Ran `gmcp.Room.Info` remotely from VS Code chat interface and saw results display in Mudlet console

**What Works Great:**
✅ Commands execute successfully in Mudlet  
✅ MCP server starts automatically on port 11235  
✅ External AI tools can send Lua commands  
✅ Real-time game data access (GMCP, etc.)

**Current Limitations (Future Enhancement Ideas):**
❌ **No return values** - Can't see command output back in VS Code  
❌ **No success/failure feedback** - Must check Mudlet console to verify results  
❌ **No captured console output** - Results displayed in Mudlet aren't returned to MCP client  
❌ **Limited error reporting** - Only basic `{"isError": false}` responses

**Suggestion for Enhanced AI Assistance:**
For truly interactive AI scripting, it would be valuable to capture and return:
- Lua function return values
- Console output from print() statements  
- Execution results and error messages
- GMCP/game data responses

This would enable AI assistants to see what happened and provide better guidance. Right now it's "fire and forget" - commands work but no feedback loop.

Overall: **Excellent foundation!** The MCP server successfully enables remote Mudlet control. Adding return value capture would make it perfect for AI-assisted scripting.

ZookaOnGit commented:
note to add qt-httpserver-dev, qt6-websockets-dev to the docs when merged

vadi2 committed:
Add unit tests for MCP server and Lua bridge

vadi2 committed:
Use forward declarations for Host in MCP headers

Avoids dragging in QTextDecoder dependency via Host.h -> TCommandLine.h

vadi2 committed:
Merge branch 'development' into add-mcp-server

vadi2 committed:
Remove unused QTextDecoder include from TCommandLine.h

vadi2 committed:
Fix MCP tests to link against mudlet_core with sanitizer support

vadi2 committed:
Fix MCP Lua return value capture using lua_pcall

luaL_dostring uses lua_pcall with 0 return values, so return values
were never captured. Now using luaL_loadstring + lua_pcall with
LUA_MULTRET to properly capture all return values from executed code.

Also improves error handling with separate syntax vs runtime errors.

vadi2 committed:
Merge latest development

ZookaOnGit committed:
Merge branch 'development' into add-mcp-server

vadi2 commented:
Putting this on hold until there's more interest from the community, tracking it in https://discord.com/channels/283581582550237184/792073945922142259/1479106751503798465.

vadi2 committed:
Merge branch 'development' into add-mcp-server

Assisted-by: Claude:claude-opus-4-8

vadi2 committed:
Fix: install qthttpserver for MCP server PR builds

The MCP server feature requires Qt6::HttpServer, but the PR build workflow's install-qt-action modules list omitted qthttpserver, so CMake configure failed. Mirror the release workflow which already lists it.

Assisted-by: Claude:claude-opus-4-8

vadi2 committed:
Fix: also install qtwebsockets for MCP server PR builds

Qt6HttpServer's CMake config requires Qt6WebSockets via find_dependency, so find_package(HttpServer) fails without it. The release workflow already lists both modules.

Assisted-by: Claude:claude-opus-4-8

vadi2 committed:
Fix: make INCLUDE_MCPSERVER a PUBLIC compile definition

INCLUDE_MCPSERVER gates the mpMCPServer member in the public Host.h header but was defined PRIVATE on mudlet_core. Test executables that include Host.h without the macro saw a smaller Host than the library, an ODR violation that corrupted memory and segfaulted 6 functional tests. Match INCLUDE_3DMAPPER and make it PUBLIC so every consumer agrees on the class layout.

Assisted-by: Claude:claude-opus-4-8

vadi2 committed:
Merge remote-tracking branch 'origin/development' into add-mcp-server

Assisted-by: Claude:claude-opus-4-8

vadi2 committed:
Merge remote-tracking branch 'origin/development' into add-mcp-server

Assisted-by: Claude:claude-opus-4-8

vadi2 commented:
Should be updated to support stateless MCP: https://blog.modelcontextprotocol.io/posts/2026-07-28/

vadi2 committed:
Merge remote-tracking branch 'origin/development' into add-mcp-server

Conflicts:
- src/Host.cpp: development dropped mpDockableMapWidget from the Host ctor
  init list (it moved onto TConsole in #9513); kept the MCP server member
  init and dropped the stale entry.
- test/CMakeLists.txt: both sides appended a block; kept both, and gated the
  MCP tests on USE_MCPSERVER rather than Qt6HttpServer_FOUND so they match the
  option that actually compiles the sources into the Mudlet library.

Assisted-by: Claude:claude-opus-5[1m]

vadi2 committed:
Merge remote-tracking branch 'origin/development' into add-mcp-server

vadi2 commented on src/TMCPLuaBridge.cpp:
```diff
@@ -0,0 +1,513 @@
+{
+    if (smExecutionTimer.isValid() && smExecutionTimer.elapsed() >= smExecutionDeadline) {
+        //: Error shown to an AI model when the Lua code it sent ran for too long and was stopped. %1 is a whole number of seconds.
+        luaL_error(L, "%s", TMCPLuaBridge::tr("stopped after running for more than %1 seconds").arg((smDeadlineMs + 999) / 1000).toUtf8().constData());
```
Confirmed, and reproduced rather than reasoned about - but it is a documented limitation with no fix available on Lua 5.1, so it is staying.

Reproducing the exact mechanism (`LUA_MASKCOUNT` every 20000 instructions, hook raises once a wall-clock deadline has passed):

```
plain infinite loop                stopped=true  elapsed=0.50s  hook fired 1 time(s)
loop that pcalls away the error    stopped=false elapsed=3.00s  hook fired 115597 time(s)  <-- RAN TO COMPLETION
```

So yes: the plain runaway stops at the deadline, and a loop that swallows the error round each iteration runs indefinitely, the hook firing 115597 times to no effect.

The standard fix - run the snippet in a coroutine and yield out of it from the hook - does not work here. Mudlet embeds Lua 5.1, which forbids yielding from a count hook; `coroutine.yield()` inside the hook silently fails to suspend and the loop escapes:

```
Lua version: Lua 5.1
resume returned ok=true value=ESCAPED after 2.00s
RESULT: yielding from the hook did NOT stop it
```

`lua_error` is the only way out of a hook in 5.1, and `pcall` catches it by definition. Seeding a wrapped `pcall`/`xpcall` into the snippet's environment would catch the accidental case, but `_G.pcall` bypasses it in one token, so it narrows nothing against code that means it - while changing `pcall`'s semantics for every snippet, where getting the multiple-return handling wrong is a worse bug than the one being mitigated.

Worth noting the exposure is not new: any ordinary trigger, alias or timer containing `while true do end` freezes Mudlet identically, because there is one thread. The MCP server is off by default, binds loopback only, and requires a token, so this adds no reachable surface beyond what a user's own scripts already have. The code comment at the hook already states the limitation.
